### ISSUE TITLE
Future

### ISSUE THREAD

**Issue by guaraqe** [SamuelSchlesinger/Gaia#26]
Hey guys, I've been following the project for some time, and I really like the architecture. What are your goals for the future of numeric Haskell?

I'm asking that since we clearly are in need of, and there are efforts (involving you) for:
- a better algebra interface (like this one)
- better support for subcategories (like SubHask)
- a better numeric tower (like tower)

I have my own little effort into making my work life more liveable [here](https://github.com/guaraqe/constraint-classes) (basically changing from Hask endofunctors to functors from a subcategory into Hask), but it is not ideal.

Would you be interested into collaborating to make a library that fills these gaps while being Prelude compatible? This is not a trivial effort and many details would have to be discussed, but I think it would be worth it. I think that Prelude compatibility (as much as possible) is essential if we want it to have a chance to be used widely.

Thanks for your work guys!

**Comment by SamuelSchlesinger** [comment]
I'm quite interested in doing so but I sincerely feel confused about how to proceed. It seems as if compatibility with the standard Prelude essentially nails you to having completely parametric polymorphism on the structures we have nice syntax for (Monad, Functor, etc) which sort of defeats a lot of the purpose of creating a new interface, as taking advantage of this syntactic sugar and the proper operators is something I care a lot about.

I'd love to start this effort but essentially I feel like it's an open problem for a reason, and it's essentially because we already are working on getting up to speed with GHC8, and I doubt we could get traction enough that our ecosystem (which it would essentially have to be) be comparable to the inaccessible bit which are generally pretty intrinsically tied to the old categorical classes.

I've thought this through quite a bit lately and I'm thinking as soon as we get ourselves a pi constructor (dependent types), we can actually construct an interface which contains accompanying proof terms of the laws associated with each class, guaranteeing the correctness of implementations. If we implement a desirable DSL for writing proofs and we implement a whole bunch of the old functionality in a generalized and verified way, I think there's a crowd that would come over and be really excited to work in that new system. This should be possible by 2020 at the latest.

**Comment by SamuelSchlesinger** [comment]
So we would have something like the following:

``` Haskell

class Associative m where
  (+) :: m -> m -> m
  associates :: pi (x :: m) (y :: m) (z :: m). (x + y) + z = x + (y + z)

```

Edit: God it'd be cool if they inferred those pi variables and we could just write 

```

associates :: (x + y) + z = x + (y + z)

```

**Comment by guaraqe** [comment]
Indeed, I'm also confused on how to proceed exactly. I'm more interested on "the best we can have now", instead of necessarily the best possible interface.

My concept of `Prelude` compatibility is kinda large, basically no conflict of names and some induced instances, so that something that is defined as `Functor` in a package can be automatically a `NewFuctor Hask Hask`, with no boilerplate necessary to use the library. And also the transition should be easy, if you know a function name in regular Haskell, you should be able to deduce the same function name in the expanded framework, like `fmap -> _fmap`, `<*> -> <*>:`, etc... This is not my dream neither, but I have no better idea.

The syntactic sugar part is not so bad I think, as the most important one is `do` notation and not having it for array-like structures is not that much a loss.

For the laws, I like SubHask's system's of automatically writing tests for the laws. Seems to be a good middle point. But constraint induction dictionaries are nice too.

What are the things that you have more interest in, or you need the most? I think that if we have a nice list of "possible wishes", that will enlighten the path. For me:
- I work with biophysics, so I have to do a ton of numerical calculations, and I have some strange algebraic structures that I work with. So the subcategory thing for me is a must.
- I would love to unify the category and algebra interface with "a monoid/semigroup is a category/semicategory with only one element" thing. And also the stuff from `mono-traversable` using single point constraints. And `contravariant` stuff using opposite categories.
- Separate some type classes in more logical ways. For example, in SubHask not every `Monad` is not an `Applicative` simply because in Haskell `Applicative` supposes a cartesian closed category and most constraints are not cartesian closed. I played with the idea of separating objects and homs [here](https://github.com/guaraqe/constraint-categories/blob/master/src/Category.hs). Also, `Map k` is not `Applicative` just because the class mixes lax-monoidal functors and tensorial strengh. There is a perfectly ok `<*>` (`Map.intersectWith ($)`), but no `pure`.

**Comment by SamuelSchlesinger** [comment]
I'm down to put something like that for sure, I've been using Python for my work because I simply can't access the sorts of analyses I need in Haskell land yet.

I'm imagining that you could implement pure if you had an "anointed point" or something, but I assume there's a problem with that? Other than it being arbitrary of course.

What do you mean exactly when you say subcategories like what facilities exactly do you want? 

I like the second wish you pointed out, I think that'd be a blast. I definitely agree with the third wish, that's what gaia explored.

**Comment by guaraqe** [comment]
Yeah, the Python thing is sad... But we agree a lot, that's great!

I always saw `Map k a` as `k -> Maybe a`, and so pure would need an infinite map in general... Maybe that's possible with other interpretations.

Subcategories in the sense of being able to use type classes like Functor on constraint domains, namely `Storable` when dealing with C bindings.

So the central point would be the design of the `Category` class that would allow all of this. Does the design here in gaia allows that? I think the one in the link in the last post allows all of that, but every line of code is a battle against the compiler...

**Comment by SamuelSchlesinger** [comment]
Hmm so you want constrained categories? I dunno what's "sub" about them, they're distinct categories from Hask.

We need to work on that compiler thing because I've come across similar problems. There's one GHC ticket out there that would fix a lot of problems for me, and I'll explain the problem here:

I want to do the whole kind indexed Hom, but in doing so you run into the following scenario:

``` Haskell

data VectorSpace = V Type Nat

data Linear :: VectorSpace -> VectorSpace -> Type where
    Matrix :: (Z `Mod` m -> Z `mod` n -> f) -> Linear (V f m) (V f n)

```

and then you want to write an instance for Category:

``` Haskell

instance Category m where
  id = ...

```

and you can't match on the V constructors to let you use the Linear constructor and life is ruined. Apparently this is beyond the capacity of Core at this time which made me stop trying, and there seems to be a ticket for this, though with low priority when I checked.

**Comment by guaraqe** [comment]
Well, both objects and morphisms are in Hask, so it's a subcategory.

For the Linear case, if the problem is in Core, that can take a really long time... But this encoding is ambitious too, as it tries to have typed dimensions. If you agree to have newtyped ints to represent known spaces, the problem becomes simpler...

The full polykinded Hom is not the only possible way to have different categories in Hask. Perhaps a less powerful encoding would bring a better equilibrium. That's why I think that looking for the most suitable design for `Category` is central.

What do you think about making a repo listing the things we expect of the library, take some time to bring some candidates to the table, listing the possibilities of each one and decide the one that seems to be the best compromise?

**Comment by SamuelSchlesinger** [comment]
Yeah, you should bring this discussion to DataHaskell: https://gitter.im/dataHaskell/Lobby

**Comment by guaraqe** [comment]
Seems to be the best, indeed. I'll write something more organized and come back there.

**Comment by tonyday567** [comment]
Nice thread. dataHaskell is awesome but I like github issues as well, for a bit more permanence. 

There's plenty to keep me interested in endo haskell, and I take the compiler fighting as a sign saying 'too soon'. So that leaves me more in the camp of what we could do right now: algebraic structure like gaia without coerce (yes, it will be ugly), newtype Ints for size as you say, and look to fit in with the emerging ML ecosystem.
Ultimately, I personally would just like Haskell numerical code to look more like math, and be performant.
My gut feel is that this would be enough for algorithm devs to stay sympathetic to ghc and enjoy the experience.
I would enjoy collaborating and think starting a repo setting out design issues, exploring tradeoffs etc is an excellent idea, even if it doesn't end up having longevity.
I don't know enough about ML though, so I'm planning on spending a fair bit of time on learning and writing algos, along the lines of recent dataHaskell discussions favoring clarity.
